YakimS/assignment2/BinaryTree.java

```diff
@@ -124,108 +71,83 @@ else if(printAncestors(key,root.getLeftSon()) ||
+    public T getLowestCommonAncestor(T value1, T value2) throws IllegalArgumentException {
 
-        // create paths to the two keys given
-        List<Integer> firstKeyPath = getKeyPath(key1, root, new LinkedList<>());
-        List<Integer> secKeyPath = getKeyPath(key2, root, new LinkedList<>());
+        TreeNode<T> nodeInPathOfVal1 = findNodeByVal(value1, root);
+        TreeNode<T> nodeInPathOfVal2 = findNodeByVal(value2, root);
 
-        if(firstKeyPath == null || secKeyPath == null)
+        if(nodeInPathOfVal1 == null || nodeInPathOfVal2 == null)
         {
-            throw new KeyNotFoundException();
+            throw new IllegalArgumentException("Value not found");
+        }
+        else if(nodeInPathOfVal1.equals(nodeInPathOfVal2))
+        {
+            return nodeInPathOfVal1.getValue();
+        }
+        else if(nodeInPathOfVal1.equals(root) || nodeInPathOfVal2.equals(root))
+        {
+            return root.getValue();
         }
 
-        Collections.reverse(firstKeyPath);
-        Collections.reverse(secKeyPath);
-
-        // iterate throw the paths till they are stop intersecting - this will
-        // be where the lowest common ancestor at
-        int currentLowestCommonAncestor = root.key;
-        Iterator IteratorFirst = firstKeyPath.iterator();
-        Iterator IteratorSecond = secKeyPath.iterator();
-        while (IteratorFirst.hasNext() && IteratorSecond.hasNext()) {
-            int nextInFirstPath = (int) IteratorFirst.next();
-            int nextInSecondPath = (int) IteratorSecond.next();
-            if(nextInFirstPath == nextInSecondPath)
+        while (nodeInPathOfVal1.getLevel() != 0 && nodeInPathOfVal2.getLevel()!=0)
+        {
+            if(nodeInPathOfVal1.equals(nodeInPathOfVal2))
+            {
+                return nodeInPathOfVal1.getValue();
+            }
+            else if(nodeInPathOfVal1.getLevel() < nodeInPathOfVal2.getLevel())
+            {
+                nodeInPathOfVal2 = nodeInPathOfVal2.getParent();
+            }
+            else if(nodeInPathOfVal1.getLevel() > nodeInPathOfVal2.getLevel())
             {
-                currentLowestCommonAncestor = nextInFirstPath;
+                nodeInPathOfVal1 = nodeInPathOfVal1.getParent();
             }
-            else
+            else if (nodeInPathOfVal1.getLevel() == nodeInPathOfVal2.getLevel())
             {
-                break;
+                nodeInPathOfVal1 = nodeInPathOfVal1.getParent();
+                nodeInPathOfVal2 = nodeInPathOfVal2.getParent();
             }
         }
-        return currentLowestCommonAncestor;
+
+        return null;
     }
-    private LinkedList<Integer> getKeyPath(int key, Node root, LinkedList<Integer> linkedList)
+
+    private TreeNode<T> findNodeByVal(T value1, TreeNode<T> root)
     {
         if(root == null)
         {
             return null;
         }
-        if(root.getKey() == key)
+        else if(root.getValue() == value1)

```

Comment:
This should be `.equals`.  Two `Integer` (or `String`, or ...) objects can carry the same value, but not be equal when compared via `==`.